### PR TITLE
Refactor exception handling

### PR DESCRIPTION
**Pull request by nilsthamm**
The code changed by this PR will mostly be invalid as soon as #882  is finished. So the changes in the SQLTransaltor are to get feedback on what we want to declare as invalid input and how we want to to the exception handling conceptionally.

#545 and #830 both required a new way of exception handling.
On the one hand, exceptions in tasks will now be caught and propagated to the caller that scheduled the task. New is that this now also works when the task got executed in a worker with a scheduler in place. (Mostly changes from [mrks/scheduling-throw](https://github.com/hyrise/hyrise/tree/mrks/scheduling-throw))

The Exception InvalidInput is to mark an error that was most likely caused by wrong user input and that will be caught by the console. Therefore it doesn't causes a termination and a rollback of the transaction will be attempted. This also means all other types of exceptions will cause the console to crash

### CODE REVIEW

File: src/lib/sql/sql_pipeline_statement.cpp [hunk 1]
```diff
@@ -159,7 +156,9 @@ const std::shared_ptr<SQLQueryPlan>& SQLPipelineStatement::get_query_plan() {
     Assert(_prepared_statements, "Cannot execute statement without prepared statement cache.");
     const auto plan = _prepared_statements->try_get(execute_statement->name);
 
-    Assert(plan, "Requested prepared statement does not exist!");
+    if (!plan) {
```
Comment (lawben) [inline src/lib/sql/sql_pipeline_statement.cpp]: These one-line `if () throw` statements should probably be turned into `AssertInput`

**Comment by mrks** [comment]
Is this still WIP?

File: src/bin/console/console.cpp [hunk 2]
```diff
@@ -242,7 +243,7 @@ bool Console::_initialize_pipeline(const std::string& sql) {
       _sql_pipeline = std::make_unique<SQLPipeline>(
           SQLPipelineBuilder{sql}.with_prepared_statement_cache(_prepared_statements).create_pipeline());
     }
-  } catch (const std::exception& exception) {
+  } catch (const InvalidInput& exception) {
```
Comment (mrks) [inline src/bin/console/console.cpp]: I'd rather call this `InvalidInputException` so that its purpose is clear

File: src/lib/sql/sql_pipeline_statement.cpp [hunk 3]
```diff
@@ -51,14 +51,11 @@ const std::shared_ptr<hsql::SQLParserResult>& SQLPipelineStatement::get_parsed_s
...
-    hsql::SQLParser::parse(_sql_string, _parsed_sql_statement.get());
-  } catch (const std::exception& exception) {
-    throw std::runtime_error("Error while parsing SQL query:\n  " + std::string(exception.what()));
-  }
+
+  hsql::SQLParser::parse(_sql_string, _parsed_sql_statement.get());
 
   if (!_parsed_sql_statement->isValid()) {
-    throw std::runtime_error(SQLPipelineStatement::create_parse_error_message(_sql_string, *_parsed_sql_statement));
+    throw InvalidInput(SQLPipelineStatement::create_parse_error_message(_sql_string, *_parsed_sql_statement));
```
Comment (mrks) [inline src/lib/sql/sql_pipeline_statement.cpp]: `AssertInput`? Other places as well.

File: src/lib/sql/sql_translator.cpp [hunk 4]
```diff
@@ -134,15 +134,17 @@ std::shared_ptr<AbstractLQPNode> SQLTranslator::translate_statement(const hsql::
...
   auto target_table = StorageManager::get().get_table(table_name);
 
-  Assert(target_table != nullptr, "Insert: Invalid table name");
 
   std::shared_ptr<AbstractLQPNode> current_result_node;
 
   // Check for SELECT ... INTO .. query
   if (insert.type == hsql::kInsertSelect) {
-    DebugAssert(insert.select != nullptr, "Insert: no select statement given");
+    AssertInput(insert.select != nullptr, std::string("Insert: no select statement given"));
```
Comment (mrks) [inline src/lib/sql/sql_translator.cpp]: Why do we have to create an `std::string` here?

**Comment by nilsthamm** [comment]
Depends on if we want to wait for #882. When I created it, I counted on a merge in near time, but I can also create a second PR after Moritz PR is merged.

File: src/lib/scheduler/abstract_task.cpp [hunk 5]
```diff
@@ -81,20 +89,48 @@ void AbstractTask::join() {
...
 }
 
 void AbstractTask::execute() {
   DebugAssert(!(_started.exchange(true)), "Possible bug: Trying to execute the same task twice");
   DebugAssert(is_ready(), "Task must not be executed before its dependencies are done");
 
-  _on_execute();
+  try {
+    _on_execute();
+  } catch (...) {
```
Comment (mrks) [inline src/lib/scheduler/abstract_task.cpp]: > * don't catch "serious" issues, only invalid_argument
> * serious issues (e.g., (Debug)Assert should still lead to std::terminate

Comment (nilsthamm) [reply src/lib/scheduler/abstract_task.cpp]: My understanding was, that we want to catch everything here (so that the worker won't die). Then we notify the caller or if there is no worker the throw the exception again. I thought the "don't catch "serious" issues, only invalid_argument" part applied to the specific places where the handle user input

Comment (mrks) [reply src/lib/scheduler/abstract_task.cpp]: No - if we have an actual error, we still want to die as early as possible and not catch them. This is only about exceptions that are caused by user errors (e.g., mismatching data types for insert). *That* is something that the database should survive.

File: src/lib/sql/sql_translator.cpp (hunk 4 above)
Comment (nilsthamm) [reply src/lib/sql/sql_translator.cpp]: Otherwise, it will be handled as const char* and we can't add the "Invalid input error: " extension to the output message in src/lib/utils/assert.hpp . Or do you think we don't need to add the extension every time and it should be  part of the initial string?

File: src/lib/sql/sql_pipeline.cpp [hunk 6]
```diff
@@ -15,14 +16,11 @@ SQLPipeline::SQLPipeline(const std::string& sql, std::shared_ptr<TransactionCont
...
-    hsql::SQLParser::parse(sql, &parse_result);
-  } catch (const std::exception& exception) {
-    throw std::runtime_error("Error while parsing SQL query:\n  " + std::string(exception.what()));
-  }
+
+  hsql::SQLParser::parse(sql, &parse_result);
 
   if (!parse_result.isValid()) {
-    throw std::runtime_error(SQLPipelineStatement::create_parse_error_message(sql, parse_result));
+    throw InvalidInput(SQLPipelineStatement::create_parse_error_message(sql, parse_result));
```
Comment (mrzzzrm) [inline src/lib/sql/sql_pipeline.cpp]: Should be AssertInput

File: src/lib/sql/sql_translator.cpp (hunk 4 above)
Comment (mrzzzrm) [reply src/lib/sql/sql_translator.cpp]: AssertInput should be able to take a string literal, pls do all necessary conversions/use std::stringstream in AssertInput.

File: src/lib/utils/exception.hpp [hunk 7]
```diff
@@ -0,0 +1,16 @@
+#pragma once
+
+#include <iostream>
+#include <exception>
+#include <stdexcept>
+#include <sstream>
+
+namespace opossum {
+
+class InvalidInput : public std::runtime_error {
```
Comment (mrzzzrm) [inline src/lib/utils/exception.hpp]: I appreciate that InvalidInput is small, but please use the same name for the file anyway. Think of the people who look for classes using their file browser.

File: src/lib/scheduler/abstract_task.cpp (hunk 5 above)
Comment (nilsthamm) [reply src/lib/scheduler/abstract_task.cpp]: Ok, I will change this

File: src/lib/utils/exception.hpp (hunk 7 above)
Comment (mrzzzrm) [reply src/lib/utils/exception.hpp]: also pls doc that this is used by AssertInput

Comment (nilsthamm) [reply src/lib/utils/exception.hpp]: Thought this could be a file like assert.hpp, where we have all exceptions in one place (since they are usually very short). Or do you think we don't need this?

**Comment by nilsthamm** [comment]
Thanks for the feedback, I incorporated all changes. I'm now waiting for the rest of the review from @mrzzzrm and will then merge if no further changes are requested

File: src/lib/utils/exception.hpp (hunk 7 above)
Comment (mrzzzrm) [reply src/lib/utils/exception.hpp]: Don't get me wrong, but things that *could* be something if *someone in the future* does it, rarely turn out to actually *be* that particular something :)

We can think about a unified exception compile unit if there are actually multiple hyrise-defined exceptions, until then its just a normal class with a normal file name

**Comment by mrzzzrm** [comment]
As said, can look at it during the weekend, but won't block merging if @mrks is happy

**Comment by mrks** [comment]
@mrzzzrm - how to we proceed?

File: src/lib/scheduler/abstract_task.cpp [hunk 8]
```diff
@@ -81,20 +90,48 @@ void AbstractTask::join() {
 void AbstractTask::_join_without_replacement_worker() {
   std::unique_lock<std::mutex> lock(_done_mutex);
   _done_condition_variable.wait(lock, [&]() { return _done; });
+
+  // An exception was thrown in the task. We caught it there so that the worker does not crash. This might be an
+  // exception from a predecessor task. If you get an exception here, a good idea is to disable the scheduler so
+  // that it throws immediately. Also, you could use gcc with "catch throw"
```
Comment (mrzzzrm) [inline src/lib/scheduler/abstract_task.cpp]: I suppose you mean gdb and not gcc?

File: src/lib/utils/exception.hpp [hunk 9]
```diff
@@ -0,0 +1,22 @@
+#pragma once
```
Comment (mrzzzrm) [inline src/lib/utils/exception.hpp]: I stand by my point that this file should be called invalid_input_exception.hpp
Also, the cpp is empty and not needed.

File: src/test/scheduler/scheduler_test.cpp [hunk 10]
```diff
@@ -164,23 +166,109 @@ TEST_F(SchedulerTest, DiamondDependenciesWithScheduler) {
 }
 
 TEST_F(SchedulerTest, LinearDependenciesWithoutScheduler) {
+  CurrentScheduler::set(nullptr);
```
Comment (mrzzzrm) [inline src/test/scheduler/scheduler_test.cpp]: Why is this necessary now and wasn't before. If you need to do it for every test, consider doing it in  SchedulerTest::SetUp().

File: src/test/scheduler/scheduler_test.cpp [hunk 11]
```diff
@@ -164,23 +166,109 @@ TEST_F(SchedulerTest, DiamondDependenciesWithScheduler) {
...
+
+  std::vector<std::shared_ptr<AbstractTask>> jobs;
+  for (auto i = 0u; i < ProcessingUnit::MAX_WORKERS_PER_CORE + 5; ++i) {
+    auto job = std::make_shared<JobTask>([]() { throw InvalidInputException("Invalid Input!"); });
+    job->schedule();
+    jobs.emplace_back(job);
+  }
+  EXPECT_THROW(CurrentScheduler::wait_for_tasks(jobs), InvalidInputException);
+
+  bool is_there_still_a_worker = false;
```
Comment (mrzzzrm) [inline src/test/scheduler/scheduler_test.cpp]: minor: `auto` for consistency

File: src/test/scheduler/scheduler_test.cpp [hunk 12]
```diff
@@ -164,23 +166,109 @@ TEST_F(SchedulerTest, DiamondDependenciesWithScheduler) {
...
+  jobs = {};
+  auto job = std::make_shared<JobTask>([&]() { is_there_still_a_worker = true; });
+  job->schedule();
+  jobs.emplace_back(job);
+  CurrentScheduler::wait_for_tasks(jobs);
```
Comment (mrzzzrm) [inline src/test/scheduler/scheduler_test.cpp]: `schedule_and_wait_for_tasks` would be a oneliner

File: src/test/scheduler/scheduler_test.cpp [hunk 13]
```diff
@@ -164,23 +166,109 @@ TEST_F(SchedulerTest, DiamondDependenciesWithScheduler) {
...
+  int step = 0;
+  jobs.emplace_back(std::make_shared<JobTask>([&]() { step = 1; }));
+  jobs.emplace_back(std::make_shared<JobTask>([&]() {
+    step = 2;
+    throw InvalidInputException("Boom");
+  }));
+  jobs.emplace_back(std::make_shared<JobTask>([&]() { step = 3; }));
+  jobs[0]->set_as_predecessor_of(jobs[1]);
+  jobs[1]->set_as_predecessor_of(jobs[2]);
+  jobs[2]->schedule();
```
Comment (mrzzzrm) [inline src/test/scheduler/scheduler_test.cpp]: again, schedule_and_wait_for_tasks would be shorter

File: src/test/scheduler/scheduler_test.cpp [hunk 14]
```diff
@@ -164,23 +166,109 @@ TEST_F(SchedulerTest, DiamondDependenciesWithScheduler) {
...
+  jobs[1]->schedule();
```
Comment (mrzzzrm) [inline src/test/scheduler/scheduler_test.cpp]: schedule_and_wait_for_tasks here as well

File: src/bin/console/console.cpp [hunk 15]
```diff
@@ -36,6 +36,7 @@
 #include "sql/sql_translator.hpp"
 #include "storage/storage_manager.hpp"
 #include "tpcc/tpcc_table_generator.hpp"
+#include "utils/exception.hpp"
 #include "utils/filesystem.hpp"
 #include "utils/load_table.hpp"
```
Comment (mrzzzrm) [inline src/bin/console/console.cpp]: random line: why update cxxopts?

File: src/lib/scheduler/abstract_task.cpp [hunk 16]
```diff
@@ -81,20 +90,48 @@ void AbstractTask::join() {
...
   DebugAssert(!(_started.exchange(true)), "Possible bug: Trying to execute the same task twice");
   DebugAssert(is_ready(), "Task must not be executed before its dependencies are done");
 
-  _on_execute();
+  try {
+    _on_execute();
+  } catch (InvalidInputException &e) {
+    if (_exception_callback) {
+      // An exception callback was explicitly provided so we let the receiver handle the exception
+      _exception_callback(std::current_exception());
```
Comment (mrzzzrm) [inline src/lib/scheduler/abstract_task.cpp]: mark_as_done() is not called here, why? Someone might be waiting for the task and will now lock forever.

File: src/lib/scheduler/abstract_task.cpp [hunk 17]
```diff
@@ -81,20 +90,48 @@ void AbstractTask::join() {
...
+      // We don't want the worker to die because of an exception. Instead, whoever created the task should deal with it
+      _exception = std::current_exception();
+
+      _mark_as_done();
+      for (auto& successor : _successors) {
+        successor->_on_exception_in_predecessor(_exception);
+      }
+    } else {
+      // No worker, so we just throw it right here and now.
+      std::rethrow_exception(std::current_exception());
```
Comment (mrzzzrm) [inline src/lib/scheduler/abstract_task.cpp]: No mark_as_done() here either, see above?

File: src/lib/scheduler/abstract_task.cpp [hunk 18]
```diff
@@ -108,6 +145,20 @@ void AbstractTask::_mark_as_scheduled() {
...
+  _exception = exception;
+  _mark_as_done();
+  for (auto& successor : _successors) {
+    successor->_on_exception_in_predecessor(exception);
+  }
+
+  if (CurrentScheduler::is_set()) {
+    // update number of tasks executed so that Scheduler::finish finds the correct number of executed threads
+    auto worker = Worker::get_this_thread_worker();
+    worker->processing_unit().lock()->on_worker_finished_task();
```
Comment (mrzzzrm) [inline src/lib/scheduler/abstract_task.cpp]: For diamond shapes, this will lead to `on_worker_finished_task()` getting called multiple times for a single task.

**Comment by mrzzzrm** [comment]
Sorry to be so critical here. But I believe the only thing we can get such a hard-to-test component like the scheduler stuff to be stable is by keeping it simple - and by simple I mean really, really simple. As few loc as possible. Any additional services like dealing with exceptions in a flexible way just makes the number of possible, implicit states explode and then we can't reason about the scheduler anymore when it starts to act up again (let @lawben sing you a song about that)

**Comment by mrks** [comment]
Let's discuss this tomorrow.

**Comment by nilsthamm** [comment]
@mrks @mrzzzrm As discussed I removed all the changes related to the exception handling in the worker.

File: src/lib/utils/invalid_input_exception.hpp [hunk 19]
```diff
@@ -0,0 +1,22 @@
...
+#include <exception>
+#include <stdexcept>
+#include <sstream>
+
+namespace opossum {
+
+
+/*
+ * Hyrise specific exception used to handle errors related to wrong user input.
+ * The console will catch this exception and (try to) rollback related transactions.
```
Comment (mrzzzrm) [inline src/lib/utils/invalid_input_exception.hpp]: I don't think we do a rollback in the sense of a mvcc-rollback, do we? if we don't then this comment would be confusing

Comment (nilsthamm) [reply src/lib/utils/invalid_input_exception.hpp]: Good point. I misinterpreted the function call `_handle_rollback` in the console. It basically just checks if the transaction was aborted properly and not yet committed
